apps/backend/src/app/modules/notifications/notifications.service.ts

```diff
@@ -0,0 +1,142 @@
+import { BadRequestException, Inject, Injectable } from '@nestjs/common';
+import { EXTENDED_PRISMA_SERVICE } from '../database/db.constants';
+import {
+  ExtendedPrismaService,
+  ExtendedPrismaServiceTransaction
+} from '../database/prisma.extension';
+import {
+  NotificationDto,
+  NotifsMarkAsReadQueryDto,
+  PagedResponseDto,
+  NotifsGetQueryDto
+} from '../../dto';
+import { NotificationType } from '@momentum/constants';
+import { Prisma } from '@prisma/client';
+
+export type AnnouncementData = {
+  type: NotificationType.ANNOUNCEMENT;
+  message: string;
+};
+export type WRAchievedData = {
+  type: NotificationType.WR_ACHIEVED;
+  runID: bigint;
+};
+export type MapStatusChangeData = {
+  type: NotificationType.MAP_STATUS_CHANGE;
+  mapID: number;
+};
+export type MapTestReqData = {
+  type: NotificationType.MAP_TEST_INVITE;
+  requesterID: number;
+  mapID: number;
+};
+export type ReviewPostedData = {
+  type: NotificationType.REVIEW_POSTED;
+  reviewerID: number;
+  mapID: number;
+  reviewID: number;
+};
+export type NotificationData =
+  | AnnouncementData
+  | WRAchievedData
+  | MapStatusChangeData
+  | MapTestReqData
+  | ReviewPostedData;
+
+@Injectable()
+export class NotificationsService {
+  constructor(
+    @Inject(EXTENDED_PRISMA_SERVICE) private readonly db: ExtendedPrismaService
+  ) {}
+  async getNotifications(
+    userID: number,
+    query: NotifsGetQueryDto
+  ): Promise<PagedResponseDto<NotificationDto>> {
+    // Only a couple of these fields are on a notification at a time
+    // However, the include will only fetch those fields if they exist
+    // This is faster than explicitly checking
+    const dbResponse = await this.db.notification.findManyAndCount({
+      where: { notifiedUserID: userID },
+      include: {
+        user: true,
+        map: true,
+        run: true,
+        review: true
+      },
+      skip: query.skip,
+      take: query.take
+    });
+    return new PagedResponseDto(NotificationDto, dbResponse);
+  }
+
+  async markAsRead(
+    userID: number,
+    query: NotifsMarkAsReadQueryDto
+  ): Promise<void> {
+    if (query.all) {
+      await this.db.notification.deleteMany({
+        where: {
+          notifiedUserID: userID,
+          type: { not: NotificationType.MAP_TEST_INVITE }
+        }
+      });
+    } else {
+      if (!query.notifIDs)
+        throw new BadRequestException('notifIDs required if all is false');
+      for (const notifID of query.notifIDs)
+        if (Number.isNaN(notifID))
+          throw new BadRequestException('notifIDs must contain numbers only');
```

Comment:
`IntCsvQueryProperty` is already doing this check

Reply:
IMO cleaner to throw if `(!query.all && !query.notifIDs))` then in `deleteMany` do `id: query.notifIDs ? { in: query.notifIDs } : undefined`

Reply:
You just mentioned to me that IntCsvQueryProperty isn't handling NaN right, reminder for you to take a look at that!